src/lib/hooks/swap/useBestTrade.ts

```diff
@@ -0,0 +1,103 @@
+    return [trade.outputAmount, trade.outputAmount.currency, trade.inputAmount.currency]
+  }
+  return [undefined, undefined, undefined]
+}
+
+// Checks for case where debounced values are stale compared to latest values from trade.
+function isSyncing(
+  trade: InterfaceTrade<Currency, Currency, TradeType> | undefined,
+  debouncedAmount: CurrencyAmount<Currency> | undefined,
+  debouncedOtherCurrency: Currency | undefined,
+  amountFromLatestTrade: CurrencyAmount<Currency> | undefined,
+  currencyFromTrade: Currency | undefined,
+  otherCurrencyFromTrade: Currency | undefined
+): boolean {
+  // Ensure that amount from user input matches latest trade.
+  const amountsMatch = trade && debouncedAmount && amountFromLatestTrade?.equalTo(debouncedAmount)
+
+  // Ensure active swap currencies match latest trade.
+  const currenciesMatch =
+    currencyFromTrade &&
+    otherCurrencyFromTrade &&
+    debouncedAmount?.currency?.equals(currencyFromTrade) &&
+    debouncedOtherCurrency?.equals(otherCurrencyFromTrade)
+
+  return !amountsMatch || !currenciesMatch
+}
+
+/**
+ * Returns the best v2+v3 trade for a desired swap.
+ * @param tradeType whether the swap is an exact in/out
+ * @param amountSpecified the exact amount to swap in/out
+ * @param otherCurrency the desired output/payment currency
+ */
+export function useBestTrade(
+  tradeType: TradeType,
+  amountSpecified?: CurrencyAmount<Currency>,
+  otherCurrency?: Currency
+): {
+  state: TradeState
+  trade: InterfaceTrade<Currency, Currency, TradeType> | undefined
+} {
+  // Debounce is used to prevent excessive requests to SOR, as it is data intensive.
+  // This helps provide a "syncing" state the UI can reference for loading animations.
+  const [debouncedAmount, debouncedOtherCurrency] = useDebounce(
+    useMemo(() => [amountSpecified, otherCurrency], [amountSpecified, otherCurrency]),
+    200
+  )
+
+  const clientSORTrade = useClientSideSmartOrderRouterTrade(tradeType, debouncedAmount, debouncedOtherCurrency)
+
+  const [amountFromLatestTrade, currencyFromTrade, otherCurrencyFromTrade] = getTradeInputs(
+    clientSORTrade.trade,
+    tradeType
+  )
+
+  const debouncing =
+    (amountSpecified && debouncedAmount && amountSpecified !== debouncedAmount) ||
+    (debouncedOtherCurrency && otherCurrency && debouncedOtherCurrency !== otherCurrency)
+
+  const syncing = isSyncing(
+    clientSORTrade?.trade,
+    debouncedAmount,
+    debouncedOtherCurrency,
+    amountFromLatestTrade,
+    currencyFromTrade,
+    otherCurrencyFromTrade
+  )
+
+  const useFallback = !syncing && clientSORTrade.state === TradeState.NO_ROUTE_FOUND
+
+  // Use a simple client side logic as backup if SOR is not available.
+  const fallbackTrade = useClientSideV3Trade(
+    tradeType,
+    useFallback ? debouncedAmount : undefined,
+    useFallback ? debouncedOtherCurrency : undefined
+  )
+
+  return useMemo(
+    () => ({
+      ...(useFallback ? fallbackTrade : clientSORTrade),
```

Comment:
nit: Will `fallbackTrade` be undefined if the amount/otherCurrency are not passed to it? If that's the case, I think `fallbackTrade || clientSORTrade` is clearer.

Reply:
the goal here is to only use the fallback when trades are syncing and the SOR package hasnt found a route, which is why we have the useFallback check

Theres a case where both trades are valid, but we dont wanna use fallbacktrade in that case